phntmxyz/sidekick
# Add UpdateCommand

Fixes #96 

## Scope of this PR
- `pubspec.yaml` template now contains a field `sidekick: cli_version: <version>` which contains the `sidekick` version the CLI was generated with
- Checks whether an update is available whenever a command is executed
- Add `UpdateCommand` which calls the latest available version of `update_sidekick_cli.dart`
- `update_sidekick_cli` currently updates the following:
  - shell scripts in `tool` (`download_dart.sh`, `install.sh`, `run.sh`,  `sidekick_config.sh`)
  - entrypoint executable
  - `sidekick_core` dependency to latest available version
  - `sidekick->cli_version`  in `pubspec.yaml` to latest available `sidekick` version

## Necessary work in following PRs
### Update `sidekick_core` in `sidekick` and pass new sidekick version parameter to `SidekickTemplateProperties`
When generating a new sidekick CLI package, we have to save the version of sidekick which generated the CLI in the generated CLI's `<generated-cli>/pubspec.yaml`. However, we can't read the global sidekick's version because its executable is compiled and thus does not have access to its own `<global-sidekick>/pubspec.yaml` because it isn't bundled with the compiled executable.  
Instead we could use an automatically generated source file which contains the version information
See e.g. https://stackoverflow.com/a/69048343/9905602 or how mason_cli implements this in a version.dart file: https://github.com/felangel/mason/pull/538/files

We should save the version of `sidekick_core` from `sidekick`'s `pubspec.lock` in a Dart file

### Add migration functions
Whenever additional steps (besides regenerating shell scripts) are needed to migrate from an old to the latest version, add them to `update_sidekick_cli.dart`. This script receives the current and the latest version. This information can be used to call migration functions to do additional update steps (e.g. modifying certain lines in dart files).


## How to test this PR
There is a `update_command_test.dart` which already tests this command.
If you want to test this manually, do the following:
1. Edit the `pubspec.yaml` of an old sidekick CLI

```yaml
dependencies:
  sidekick_core: '>=0.11.0 <1.0.0'

dependency_overrides:
  sidekick_core:
    path: /Users/pepe/repos/sidekick/sidekick_core
```
2. Add `SidekickCommand` to the cli
3. Run `<cli> sidekick update`
4. Verify the following:  
    1. the 4 shell scripts in `tool` have been updated  
    2. `cli_version` has been updated to the latest version of `sidekick` (`0.7.1`)  
    3.  the `sidekick_core` dependency has been updated to `sidekick_core: '>=0.11.1 <1.0.0'`
    4. A success message has been printed (`Successfully updated sidekick CLI <cli> from version 0.0.0 to 0.7.1!`)
